apps/keyboard/test/unit/settings/layout_dictionary_test.js

```diff
@@ -0,0 +1,615 @@
+
+    dbDeleteItemDeferred = new Deferred();
+    listStub.dbStore.deleteItem.returns(dbDeleteItemDeferred.promise);
+
+    layoutItem = {
+      id: 'ab',
+      updateInstallProgress: this.sinon.stub()
+    };
+
+    layoutItem2 = {
+      id: 'cd',
+      updateInstallProgress: this.sinon.stub()
+    };
+  });
+
+  suite('STATE_PRELOADED', function() {
+    setup(function() {
+      var dict = {
+        filename: 'bar.dict',
+        imEngineId: 'foo',
+        filePath: 'path/to/bar.dict',
+        fileSize: 100,
+        databaseId: 'foo/bar.dict',
+        preloaded: true,
+        installedLayoutIds: new Set(['ab'])
+      };
+
+      layoutDict = new LayoutDictionary(listStub, dict);
+      layoutDict.start();
+
+      assert.equal(layoutDict.filename, 'bar.dict');
+      assert.equal(layoutDict.imEngineId, 'foo');
+      assert.equal(layoutDict.filePath, 'path/to/bar.dict');
+      assert.equal(layoutDict.fileSize, 100);
+      assert.equal(layoutDict.databaseId, 'foo/bar.dict');
+      assert.equal(layoutDict.installedLayoutIds, dict.installedLayoutIds);
+      assert.equal(layoutDict.state, layoutDict.STATE_PRELOADED);
+    });
+
+    test('installForLayout()', function() {
+      assert.throw(layoutDict.installForLayout.bind(layoutDict));
+    });
+
+    test('removeForLayout()', function() {
+      assert.throw(layoutDict.removeForLayout.bind(layoutDict));
+    });
+  });
+
+  suite('STATE_INSTALLABLE', function() {
+    setup(function() {
+      var dict = {
+        filename: 'bar.dict',
+        imEngineId: 'foo',
+        filePath: 'path/to/bar.dict',
+        fileSize: 100,
+        databaseId: 'foo/bar.dict',
+        preloaded: false,
+        installedLayoutIds: new Set()
+      };
+
+      layoutDict = new LayoutDictionary(listStub, dict);
+      layoutDict.start();
+
+      assert.equal(layoutDict.filename, 'bar.dict');
+      assert.equal(layoutDict.imEngineId, 'foo');
+      assert.equal(layoutDict.filePath, 'path/to/bar.dict');
+      assert.equal(layoutDict.fileSize, 100);
+      assert.equal(layoutDict.databaseId, 'foo/bar.dict');
+      assert.equal(layoutDict.installedLayoutIds, dict.installedLayoutIds);
+      assert.equal(layoutDict.state, layoutDict.STATE_INSTALLABLE);
+    });
+
+    test('installForLayout(layoutItem)', function() {
+      layoutDict.installForLayout(layoutItem);
+
+      assert.equal(layoutDict.state, layoutDict.STATE_INSTALLING_CANCELLABLE);
+
+      assert.isTrue(layoutDictionaryDownloaderStub.load.calledOnce);
+
+      layoutDictionaryDownloaderStub.onprogress(10, 100);
+      assert.equal(layoutDict.downloadLoadedSize, 10);
+      assert.equal(layoutDict.downloadTotalSize, 100);
+      assert.isTrue(layoutItem.updateInstallProgress.calledWith(10, 100));
+    });
+
+    test('removeForLayout(layoutItem)', function(done) {
+      layoutItem = {
+        id: 'ab',
+        updateInstallProgress: this.sinon.stub()
+      };
+      var p = layoutDict.removeForLayout(layoutDict);
+
+      p
```

Comment:
The newline looks strange...how about this? (and a few other places in this file)

      p.then(function() {
        assert.equal(
          layoutDict.state, layoutDict.STATE_INSTALLING_CANCELLABLE);
        assert.isFalse(layoutDict.installedLayoutIds.has('ab'));

        assert.isTrue(layoutDictionaryDownloaderStub.load.calledOnce);
      })
      .then(done, done);